File: web/tests/reference/tries/word-search-ii.js
function findWords(board, words) {
  const root = new Map();
  for (const w of words) {
    let node = root;
    for (const ch of w) {
      if (!node.has(ch)) node.set(ch, new Map());
      node = node.get(ch);
    }
    node.set("$", w);
  }

  const rows = board.length;
  const cols = rows > 0 ? board[0].length : 0;
  const found = [];

  const dfs = (r, c, node) => {
    const ch = board[r][c];
    if (!node.has(ch)) return;
    const next = node.get(ch);
    if (next.has("$")) {
      found.push(next.get("$"));
      next.delete("$");
    }
    board[r][c] = "#";
    for (const [dr, dc] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
      const nr = r + dr;
      const nc = c + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && board[nr][nc] !== "#") {
        dfs(nr, nc, next);
      }
    }
    board[r][c] = ch;
  };

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) dfs(r, c, root);
  }
  return found;
}

File: web/tests/reference/tries/word-search-ii.py
def find_words(board, words):
    root = {}
    for w in words:
        node = root
        for ch in w:
            node = node.setdefault(ch, {})
        node["$"] = w

    rows, cols = len(board), len(board[0]) if board else 0
    found = []

    def dfs(r, c, node):
        ch = board[r][c]
        if ch not in node:
            return
        nxt = node[ch]
        word = nxt.pop("$", None)
        if word is not None:
            found.append(word)
        board[r][c] = "#"
        for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and board[nr][nc] != "#":
                dfs(nr, nc, nxt)
        board[r][c] = ch

    for r in range(rows):
        for c in range(cols):
            dfs(r, c, root)
    return found
